app/src/main/java/com/dessalines/thumbkey/keyboards/TOKThumbKey.kt

```diff
@@ -106,8 +106,16 @@ val KB_TOK_THUMBKEY_MAIN =
                     top = KeyC("ken ", displayText = "💪"),
                     topRight = KeyC("o ", displayText = "👋"),
                     right = KeyC("wile ", displayText = "💭"),
-                    bottomRight = KeyC(",", color = MUTED),
-                    bottom = KeyC(".", color = MUTED),
+                    bottomRight = KeyC(
+                        action = ReplaceTrailingWhitespace(", ", 2),
```

Comment:
Rather than doing this *replace trailing whitespace* hack, you could also add `no_spaces_after_puncuation` to the keyboard settings struct also.

Reply:
I made a setting option like that earlier, but dropped that idea when I started digging a little deeper; *a* consequence would be that we'd still need to use a fitting KeyAction when everything else was said and done, yeah? We could add that as an optional member of `CommitText` and cram a bunch of deleting-preceding-whitespace logic into it, then pass that option through via the manually-declared `KeyC` constructor, but I just thought this was a tad less invasive.

Maybe a decent intermediate is just adding the option to `KeyC` and grabbing the appropriate `KeyAction` for whatever was declared there. Keeps extra logic that won't need to run most of the time out of `CommitText`.

Reply:
The spacebar key specifically is what adds the extra spaces: https://github.com/dessalines/thumb-key/blob/main/app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt#L196

So you could either add a setting, or create a kotlin copy of the spacebar key, for toki pona specifically.

Also though, is no spaces after puncuation a toki pona rule anyway?

Reply:
Oh, the idea here is that trailing spaces are left *for each word that's typed* and they should be replaced. e.g. with the current release/layout try typing `lon` then `.`. You'll get `lon .`. Give it a shot.

"Replace trailing whitespace" meant that typing `lon` should produce `lon `, then typing `.` should *remove the preceding space* and leave the field reading `lon.` I let it read back 2, but if there's more than 2 spaces, I assume you meant to put them there.

With this change, you no longer need to fuss with the spacebar or backspace at all to get punctuation in the right places; just hit keys. It's an ergonomic improvment.

Reply:
Oooh okay I understand. In that case yep your current solution is good. But be sure to test that starting with a puncuation marker isn't going to crash thumbkey.

Reply:
Just did, all good!